[SUREFIRE-1293] Simplify org.apache.maven.plugin.surefire.report.TestSetRunListener by using the null object pattern

britter opened the pull request:
Added null object implementations for fields used by TestSetRunListener in order to simplify TestSetRunListener code.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/NullStatisticsReporter.java:
```diff
@@ -35,7 +36,7 @@
     public NullStatisticsReporter()
     {
-        super( FileUtils.getTempDirectory() );
+        super( FileUtils.getFile( FileUtils.getTempDirectory(), RandomStringUtils.randomAlphabetic( 24 ) ) );
```
It would be nice to have protected no-arg constructor of `StatisticsReporter` for these purposes and null variables of `File` and `RunEntryStatisticsMap`, etc. Not necessary to create new objects if unused.

britter replied:
@Tibor17 makes sense. If we change it like this, we don't need the test to make sure the null objects can be created. I'll modify the PR to incorporate protected default constructors and remove the empty tests.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/TestSetRunListener.java:
```diff
@@ -76,7 +76,7 @@ public TestSetRunListener( ConsoleReporter consoleReporter, FileReporter fileRep
                                boolean isPlainFormat, boolean briefOrPlainFormat )
     {
-        this.consoleReporter = consoleReporter;
+        this.consoleReporter = consoleReporter != null ? consoleReporter : new NullConsoleReporter();
```
All of these classes are in `surefire-common` project. Therefor it would be nice not to create a new object 
`new NullConsoleReporter()` but instead keep it as static final constant in abstract Factory in surefire-common.
This would mean that such null-check may be embedded in the factory, e.g. `buildConsoleReporter()`. I would say that the caller should pass non-null consistent object into this class `TestSetRunListener` via constructor and the called class should not make any guess about it (means if == null => guess hide the error). If we accepted such paradigm then all such occurrences in this PR would have to change it.

britter replied:
> All of these classes are in surefire-common project. Therefore it would be nice not to create a new object new NullConsoleReporter() but instead keep it as static final constant in abstract Factory in surefire-common.

Agreed, I can change this.

> I would say that the caller should pass non-null consistent object into this class TestSetRunListener via constructor

Yes it would certainly be better to enforce this kinds of invariants. We can do it like this, but then the constructor of `TestSetRunListener` has to check for null parameters and throw an exception if null is passed.

> If we accepted such paradigm then all such occurrences in this PR would have to change it.

I don't understand. Change the whole surefire code base for all occurrences of this pattern?

Tibor17 replied:
no no, I meant only the scope of this PR. Not whole Surefire of course.

Tibor17 replied:
@britter 
>>does this belong into DefaultReporterFactory.createReporter()
If you mean this code:
`ConsoleReporter consoleReporter = shouldReportToConsole() ? new ConsoleReporter( consoleLogger ) : null;`

Then I would keep the responsibility of creating anew reporter yet in the method `createReporter()`. I am going to fix a bug exactly in this part of code after you. We can still improve the code by small incremental changes.

britter commented on maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/report/NullConsoleReporterTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Test
+    public void testConstruction() throws Exception {
+        new NullConsoleReporter();
```
The null implementations extend some existing classes which forces me to pass values to the super constructor. If the super constructor can not handle null input parameters a call the the default constructor of the null object will fail. To make sure the null objects can be created, I've added these tests. Makes sense?

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/NullConsoleOutputReceiver.java:
```diff
@@ -31,7 +31,13 @@
+    {
+    }
+
+    public void testSetStarting(ReportEntry reportEntry )
```
here is checkstyle error

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/NullConsoleReporter.java:
```diff
@@ -33,7 +33,10 @@
+
+    static final NullConsoleReporter INSTANCE = new NullConsoleReporter();
+
+    private NullConsoleReporter() {
```
here as well
pls run the build. it explores checkstyle erros.
Did you get your IntelliJ Idea now working with our Maven checkstyle?

Tibor17 commented:
After we are done with all commits, would you be able to safely squash them in one single commit?

britter commented:
Hello @Tibor17, I'll fix the checkstyle violations and squash everything into one single commit!

britter commented:
@Tibor17 fixed checkstyle issues, moved creation of null objects to DefaultRepoterFactory and squasched everything into one commit. Hope you like it! :-)

Tibor17 commented:
@britter 
Done

Tibor17 commented:
@britter 
The build failed [1]. Did you run a local build on your side?
[1] https://builds.apache.org/job/maven-surefire/1638/console

Tibor17 commented:
@britter 
It seems you forgot to add `ASF commons` dependency.

britter commented:
@Tibor17 I don't understand. `mvn clean compile` works. Why does this fail on Jenkins?

Tibor17 commented:
Try to clone Surefire master and just run "mvn test"

Caused by: java.lang.NoClassDefFoundError: org/apache/commons/lang3/ObjectUtils
	at org.apache.maven.plugin.surefire.report.DefaultReporterFactory.createFileReporter(DefaultReporterFactory.java:111)
	at org.apache.maven.plugin.surefire.report.DefaultReporterFactory.createReporter(DefaultReporterFactory.java:92)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:102)



On Wed, Oct 12, 2016 at 2:57 PM, Benedikt Ritter <notifications@github.com>
wrote:

> @Tibor17 <https://github.com/Tibor17> I don't understand. mvn clean
> compile works. Why does this fail on Jenkins?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/maven-surefire/pull/127#issuecomment-253205875>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_yR1I15EmeNcuXoehEwBCwAB3Al1vwks5qzNk6gaJpZM4KSBgs>
> .
>



-- 
Cheers
Tibor

Tibor17 commented:
@britter 
After using one integration tests only
`<include>org/apache/**/JUnit47ParallelIT.java</include>`
I got same errors as in Jenkins.
I think the dependency `commons-lang3` is not propagated to the forked jvm. Maybe somewhere some scope but not in our code.
If you run the test and go to `target/surefire` you will see a jar file. Download it and lookup `commons` in manifest.

Tibor17 commented:
I fixed this issue where I removed [1]. The reason behind is the `IsolatedClassLoader` in plugin process which did not have shaded `commons-lang3` in original jar file of `surefire-common`. Because of one method in `commons-lang3` we should not pass such a big library. Instead I created `ObjectUtils` which has only few method and most of them is already in Java 7 utility class `java.util.Objects`. In Surefire on the top of Java 7 I will delete them and use Java 7 facilities.
[1] import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;

britter commented:
> I fixed this issue where I removed [1]

Thank you!

> Because of one method in commons-lang3 we should not pass such a big library.

ACK

> Instead I created ObjectUtils which has only few method and most of them is already in Java 7 utility class java.util.Objects. In Surefire on the top of Java 7 I will delete them and use Java 7 facilities.

Good Idea! Sorry for causing the problems.

Tibor17 commented:
@britter No big problem. Can you always run build locally? AFAIK your PC is running the build two hours. Normally 45 minutes in Jenkins and my PC Intel i7 4Core 3GHz SSD disk.
